### PR TITLE
include:irq: Add support for uart callback

### PR DESCRIPTION
**Pull request by mchindri**
- Change irq_register_interrupt -> irq_register_callback
- Add uart interrupt support in aducm/uart.c and adumc/irq.c

These change are needed to support ashynchrnous operations over uart

\+

Dummy wifi example

### CODE REVIEW

**Comment by mchindri** [comment]
v2
 - Major changes

File: v1.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
+//irq.h
+int32_t irq_register(struct irq_ctrl_desc *desc, uint32_t irq_id,
```
Comment (buha) [inline v1.c]: what is irq_id ? is it a unique interrupt identifier number from the interrupt controller table ?

Comment (mchindri) [reply v1.c]: yes

**Comment by mchindri** [comment]
v2
 - Make implementation for aducm for the choosen V2
 - Add wifi dummy example

File: drivers/platform/aducm3029/irq.c [hunk 2]
```diff
@@ -142,37 +129,56 @@ int32_t irq_ctrl_remove(struct irq_ctrl_desc *desc)
 	if (!desc || !desc->extra || !initialized)
 		return FAILURE;
 
-	initialized = 0;
+	/* Free external interrupts */
+	for (uint32_t i; i < NB_EXT_INTERRUPTS; i++)
```
Comment (amiclaus) [inline drivers/platform/aducm3029/irq.c]: You forgot to initialize `i`.

File: include/irq.h [hunk 3]
```diff
@@ -83,16 +104,10 @@ int32_t irq_ctrl_init(struct irq_ctrl_desc **desc,
...
- * @param irq_handler - The IRQ handler.
- * @param dev_instance - device instance.
- * @return SUCCESS in case of success, FAILURE otherwise.
- */
-int32_t irq_register(struct irq_ctrl_desc *desc, uint32_t irq_id,
-		     void (*irq_handler)(void *data), void *dev_instance);
+/* Register a callback to handle the irq events */
+int32_t irq_register_callback(struct irq_ctrl_desc *desc,
+			      uint32_t irq_id, IRQ_CALLBACK callback,
+			      void *context, void *config);
```
Comment (dbogdan) [inline include/irq.h]: int32_t irq_register(struct irq_ctrl_desc *desc, uint32_t irq_id,
	void *callback_context, void *callback_config,
	void (*irq_callback)(void *callback_context, uint32_t event, void *extra));

Comment (dbogdan) [reply include/irq.h]: irq_enable(irq_id)
irq_disable(irq_id)

**Comment by mchindri** [comment]
V3
- Use callback descriptor in register_callback
- Remove unregister_callback
- Rename:	irq_source_enable/disable -> irq_enable/disable
    		irq_conf.uart_desc -> irq_conf.uart_conf
			irq_conf.irq_mode -> irq_conf.xint_conf
- Initialize i in for

**Comment by mchindri** [comment]
v4:
 - fix asytle
 - fix doxy

**Pull request closed without merging**